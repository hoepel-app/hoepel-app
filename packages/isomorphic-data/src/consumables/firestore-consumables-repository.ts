import { collection, get, set } from 'typesaurus'
import { Observable, from } from 'rxjs'
import { map } from 'rxjs/operators'
import {
  ConsumablesRepository,
  Consumables,
  ConsumableProps,
} from '@hoepel.app/isomorphic-domain'

export class FirestoreConsumablesRepository implements ConsumablesRepository {
  private readonly collection = collection<{
    readonly consumables: readonly ConsumableProps[]
  }>('consumables')

  getForTenant(tenantId: string): Observable<Consumables> {
    return from(get(this.collection, tenantId)).pipe(
      map((result) => {
        if (result == null) {
          return Consumables.createEmpty(tenantId)
        }

        return Consumables.fromProps({
          consumables: result.data.consumables,
          tenantId,
        })
      })
    )
  }

  async put(entity: Consumables): Promise<void> {
    const { tenantId, ...data } = entity.toProps()
    await set(this.collection, entity.tenantId, {
      consumables: data.consumables,
    })
  }
}
