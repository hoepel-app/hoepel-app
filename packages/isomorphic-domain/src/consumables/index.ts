export { AddConsumableCommand } from './commands/add-consumable.command'
export { ChangeConsumablePriceCommand } from './commands/change-consumable-price.command'
export { RemoveConsumableCommand } from './commands/remove-consumable.command'
export { ConsumablesApplicationService } from './consumables.application-service'
export { ConsumablesRepository } from './consumables.repository'
export {
  Consumables,
  ConsumablesProps,
  Consumable,
  ConsumableProps,
} from './consumables'
