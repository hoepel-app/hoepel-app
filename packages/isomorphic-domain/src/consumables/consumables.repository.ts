import { Observable } from 'rxjs'
import { Consumables } from './consumables'

export type ConsumablesRepository = {
  getForTenant(tenantId: string): Observable<Consumables>
  put(entity: Consumables): Promise<void>
}
