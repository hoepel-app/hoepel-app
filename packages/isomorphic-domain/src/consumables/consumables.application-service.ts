import { ConsumablesRepository } from './consumables.repository'
import { Observable } from 'rxjs'
import { Consumables } from './consumables'
import { AddConsumableCommand } from './commands/add-consumable.command'
import { first } from 'rxjs/operators'
import { CommandResult } from '@hoepel.app/ddd-library/src'
import { ChangeConsumablePriceCommand } from './commands/change-consumable-price.command'
import { RemoveConsumableCommand } from './commands/remove-consumable.command'
import { RenameConsumableCommand } from './commands/rename-consumable.command'

export class ConsumablesApplicationService {
  constructor(private readonly repo: ConsumablesRepository) {}

  findConsumables(tenantId: string): Observable<Consumables> {
    return this.repo.getForTenant(tenantId)
  }

  async addConsumable(command: AddConsumableCommand): Promise<CommandResult> {
    const consumables = await this.findConsumables(command.tenantId)
      .pipe(first())
      .toPromise()

    if (!consumables.mayAddConsumable(command.consumable)) {
      return { status: 'rejected' }
    }

    await this.repo.put(consumables.withConsumableAdded(command.consumable))

    return { status: 'accepted' }
  }

  async changeConsumablePrice(
    command: ChangeConsumablePriceCommand
  ): Promise<CommandResult> {
    const consumables = await this.findConsumables(command.tenantId)
      .pipe(first())
      .toPromise()

    if (
      command.price.totalCents < 0 ||
      !consumables.consumableWithNameExists(command.consumableName)
    ) {
      return { status: 'rejected' }
    }

    await this.repo.put(
      consumables.withChangedPriceForConsumable(
        command.consumableName,
        command.price.totalCents
      )
    )

    return { status: 'accepted' }
  }

  async renameConsumable(
    command: RenameConsumableCommand
  ): Promise<CommandResult> {
    const consumables = await this.findConsumables(command.tenantId)
      .pipe(first())
      .toPromise()

    if (consumables.consumableWithNameExists(command.newName)) {
      return { status: 'rejected', reason: 'Consumable with new name exists' }
    }

    if (!consumables.consumableWithNameExists(command.oldName)) {
      return { status: 'rejected', reason: 'Consumable not found' }
    }

    await this.repo.put(
      consumables.withRenamedConsumable(command.oldName, command.newName)
    )

    return { status: 'accepted' }
  }

  async removeConsumable(
    command: RemoveConsumableCommand
  ): Promise<CommandResult> {
    const consumables = await this.findConsumables(command.tenantId)
      .pipe(first())
      .toPromise()

    if (!consumables.consumableWithNameExists(command.consumableName)) {
      return { status: 'rejected' }
    }

    await this.repo.put(
      consumables.withConsumableRemoved(command.consumableName)
    )

    return { status: 'accepted' }
  }
}
