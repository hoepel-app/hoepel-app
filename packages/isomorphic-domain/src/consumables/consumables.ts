import { Price } from '@hoepel.app/types'
import { Aggregate } from '@hoepel.app/ddd-library/src'

export type ConsumableProps = {
  readonly name: string
  readonly priceCents: number
}

export type ConsumablesProps = {
  readonly consumables: readonly ConsumableProps[]
  readonly tenantId: string
}

export class Consumables implements Aggregate {
  private constructor(private readonly props: ConsumablesProps) {}

  static createEmpty(tenantId: string): Consumables {
    return new Consumables({ tenantId, consumables: [] })
  }

  static fromProps(props: ConsumablesProps): Consumables {
    return new Consumables(props)
  }

  toProps(): ConsumablesProps {
    return this.props
  }

  get id(): string {
    return `${this.tenantId}-consumables`
  }

  get tenantId(): string {
    return this.props.tenantId
  }

  get names(): ReadonlySet<string> {
    return new Set(this.consumables.map((consumable) => consumable.name))
  }

  get consumables(): readonly Consumable[] {
    return this.props.consumables.map((props) => Consumable.fromProps(props))
  }

  withConsumableAdded(consumable: Consumable): Consumables {
    if (!this.mayAddConsumable(consumable)) {
      return this
    }

    return Consumables.fromProps({
      ...this.toProps(),
      consumables: [...this.toProps().consumables, consumable.toProps()],
    })
  }

  withConsumableRemoved(name: string): Consumables {
    return Consumables.fromProps({
      ...this.toProps(),
      consumables: this.toProps().consumables.filter((c) => c.name !== name),
    })
  }

  withChangedPriceForConsumable(
    name: string,
    newPriceCents: number
  ): Consumables {
    if (!this.consumableWithNameExists(name)) {
      return this
    }

    return Consumables.fromProps({
      ...this.toProps(),
      consumables: [
        ...this.toProps().consumables.filter((c) => c.name !== name),
        Consumable.create(name, newPriceCents).toProps(),
      ],
    })
  }

  withRenamedConsumable(oldName: string, newName: string): Consumables {
    const currentConsumable = this.findConsumableByName(oldName)

    if (this.consumableWithNameExists(newName) || currentConsumable == null) {
      return this
    }

    return Consumables.fromProps({
      ...this.toProps(),
      consumables: [
        ...this.toProps().consumables.filter((c) => c.name !== oldName),
        currentConsumable.withName(newName).toProps(),
      ],
    })
  }

  findConsumableByName(name: string): Consumable | null {
    return (
      this.consumables.find((consumable) => consumable.name === name) ?? null
    )
  }

  mayAddConsumable(consumable: Consumable): boolean {
    return !this.consumableWithNameExists(consumable.name)
  }

  consumableWithNameExists(name: string): boolean {
    return this.findConsumableByName(name) != null
  }
}

/**
 * Used to represent something that can be consumed or purchased, which is not an attendance.
 * E.g. drink, cookies, ...
 * Consumption of consumables is tied to a dayid/DayDate o(a specific date)
 * The same consumable can be consumed multiple times by the same child on the same day.
 */
export class Consumable {
  private constructor(private readonly props: ConsumableProps) {}

  static create(name: string, priceCents: number): Consumable {
    if (priceCents < 0) {
      throw new Error('Price in cents must be a positive number')
    }

    if (Math.round(priceCents) != priceCents) {
      throw new Error('Price in cents must be an integer')
    }

    return new Consumable({ name, priceCents })
  }

  static fromProps(props: ConsumableProps): Consumable {
    return new Consumable(props)
  }

  withName(newName: string): Consumable {
    return Consumable.fromProps({
      ...this.toProps(),
      name: newName,
    })
  }

  toProps(): ConsumableProps {
    return this.props
  }

  /**
   * Price of the consumable. May not be included in fiscal certificates (these costs are non-refundable)!
   */
  get price(): Price {
    return Price.fromCents(this.props.priceCents)
  }

  /**
   * Name of the consumable, e.g. "Big chocolate cookie"
   */
  get name(): string {
    return this.props.name
  }
}
