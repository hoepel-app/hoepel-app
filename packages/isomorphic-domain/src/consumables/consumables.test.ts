import { Consumables, Consumable } from './consumables'

describe('consumables', () => {
  it('id', () => {
    const consumables = Consumables.createEmpty(
      'my-tenant-id-here'
    ).withConsumableAdded(Consumable.create('Cookie', 250))

    expect(consumables.id).toEqual('my-tenant-id-here-consumables')
  })

  it('serializing and deserializing yields same object', () => {
    const consumables = Consumables.createEmpty('my-tenant-id-here')
      .withConsumableAdded(Consumable.create('Cookie', 250))
      .withConsumableAdded(Consumable.create('Soup', 300))

    const json = JSON.parse(JSON.stringify(consumables.toProps()))

    expect(Consumables.fromProps(json)).toEqual(consumables)
  })

  describe('consumables', () => {
    it('returns consumables in alphabetical order by name', () => {
      const consumables = Consumables.createEmpty('my-tenant-id-here')
        .withConsumableAdded(Consumable.create('Soup', 250))
        .withConsumableAdded(Consumable.create('Cookie', 300))
        .withConsumableAdded(Consumable.create('ZZZ', 300))
        .withConsumableAdded(Consumable.create('Soft drink', 300))

      expect(consumables.consumables).toMatchSnapshot()
    })
  })

  describe('namesSorted', () => {
    it('returns consumable names in alphabetical order', () => {
      const consumables = Consumables.createEmpty('my-tenant-id-here')
        .withConsumableAdded(Consumable.create('Soup', 250))
        .withConsumableAdded(Consumable.create('Cookie', 300))
        .withConsumableAdded(Consumable.create('ZZZ', 300))
        .withConsumableAdded(Consumable.create('Soft drink', 300))

      expect(consumables.namesSorted).toEqual([
        'Cookie',
        'Soft drink',
        'Soup',
        'ZZZ',
      ])
    })
  })

  describe('names', () => {
    it('returns the names of added consumables', () => {
      const consumables = Consumables.createEmpty('my-tenant-id-here')
        .withConsumableAdded(Consumable.create('Cookie', 250))
        .withConsumableAdded(Consumable.create('Soft drink', 50))
        .withConsumableAdded(Consumable.create('Soup', 300))

      expect(consumables.names).toEqual(
        new Set(['Cookie', 'Soup', 'Soft drink'])
      )
    })
  })

  describe('withConsumableAdded', () => {
    it('does not add consumable with same name twice', () => {
      const consumables = Consumables.createEmpty('my-tenant-id-here')
        .withConsumableAdded(Consumable.create('Cookie', 250))
        .withConsumableAdded(Consumable.create('Cookie', 300))

      expect(consumables.consumables).toHaveLength(1)
      expect(consumables.findConsumableByName('Cookie')).toMatchInlineSnapshot(`
        Consumable {
          "props": Object {
            "name": "Cookie",
            "priceCents": 250,
          },
        }
      `)
    })
  })

  describe('findConsumableByName', () => {
    it('finds existing consumable by name', () => {
      const consumables = Consumables.createEmpty(
        'my-tenant-id-here'
      ).withConsumableAdded(Consumable.create('Cookie', 250))

      expect(consumables.findConsumableByName('Cookie')).toMatchInlineSnapshot(`
        Consumable {
          "props": Object {
            "name": "Cookie",
            "priceCents": 250,
          },
        }
      `)
    })

    it('returns null if no consumable can be found', () => {
      const consumables = Consumables.createEmpty(
        'my-tenant-id-here'
      ).withConsumableAdded(Consumable.create('Cookie', 250))

      expect(consumables.findConsumableByName('Soup')).toBeNull()
    })
  })

  describe('withChangedPriceForConsumable', () => {
    it('changes the price of a consumable', () => {
      const consumables = Consumables.createEmpty('my-tenant-id-here')
        .withConsumableAdded(Consumable.create('Cookie', 250))
        .withConsumableAdded(Consumable.create('Soup', 300))

      expect(consumables.withChangedPriceForConsumable('Cookie', 100))
        .toMatchInlineSnapshot(`
        Consumables {
          "props": Object {
            "consumables": Array [
              Object {
                "name": "Soup",
                "priceCents": 300,
              },
              Object {
                "name": "Cookie",
                "priceCents": 100,
              },
            ],
            "tenantId": "my-tenant-id-here",
          },
        }
      `)
    })

    it('does nothing when consumable does not exist', () => {
      const consumables = Consumables.createEmpty('my-tenant-id-here')
        .withConsumableAdded(Consumable.create('Cookie', 250))
        .withConsumableAdded(Consumable.create('Soup', 300))

      expect(consumables.withChangedPriceForConsumable('Cooky', 100)).toEqual(
        consumables
      )
    })
  })

  describe('withConsumableRemoved', () => {
    it('removes a consumable', () => {
      const consumables = Consumables.createEmpty('my-tenant-id-here')
        .withConsumableAdded(Consumable.create('Cookie', 250))
        .withConsumableAdded(Consumable.create('Soup', 300))

      expect(consumables.withConsumableRemoved('Cookie'))
        .toMatchInlineSnapshot(`
        Consumables {
          "props": Object {
            "consumables": Array [
              Object {
                "name": "Soup",
                "priceCents": 300,
              },
            ],
            "tenantId": "my-tenant-id-here",
          },
        }
      `)
    })

    it('does nothing when consumable does not exist', () => {
      const consumables = Consumables.createEmpty('my-tenant-id-here')
        .withConsumableAdded(Consumable.create('Cookie', 250))
        .withConsumableAdded(Consumable.create('Soup', 300))

      expect(consumables.withConsumableRemoved('Cooky')).toEqual(consumables)
    })
  })

  describe('withConsumableRenamed', () => {
    const exampleConsumables = Consumables.createEmpty('my-tenant-id-here')
      .withConsumableAdded(Consumable.create('Cookie', 250))
      .withConsumableAdded(Consumable.create('Soft drink', 50))
      .withConsumableAdded(Consumable.create('Soup', 300))

    it('renames consumable', () => {
      expect(
        exampleConsumables.withRenamedConsumable(
          'Cookie',
          'Big Chocolate Cookie'
        )
      ).toMatchSnapshot()
    })

    it('does nothing when renaming non-existing consumable', () => {
      expect(
        exampleConsumables.withRenamedConsumable('Blah', 'Something')
      ).toEqual(exampleConsumables)
    })

    it('does nothing when renaming consumable to name already in use', () => {
      expect(
        exampleConsumables.withRenamedConsumable('Cookie', 'Soup')
      ).toEqual(exampleConsumables)
    })
  })

  describe('isEmpty', () => {
    it('true when no consumables added', () => {
      expect(Consumables.createEmpty('my-tenant-id').isEmpty).toEqual(true)
    })

    it('false when there are consumables added', () => {
      const consumables = Consumables.createEmpty('my-tenant-id-here')
        .withConsumableAdded(Consumable.create('Cookie', 250))
        .withConsumableAdded(Consumable.create('Soup', 300))

      expect(consumables.isEmpty).toEqual(false)
    })
  })
})

describe('consumable', () => {
  describe('create', () => {
    it('allows creating free consumable', () => {
      const consumable = Consumable.create('Water', 0)

      expect(consumable.price.totalCents).toEqual(0)
    })

    it('throws when creating with negative price', () => {
      expect(() =>
        Consumable.create('Water', -100)
      ).toThrowErrorMatchingInlineSnapshot(
        `"Price in cents must be a positive number"`
      )
    })

    it('throws when creating with floating point price', () => {
      expect(() =>
        Consumable.create('Water', 66.66)
      ).toThrowErrorMatchingInlineSnapshot(
        `"Price in cents must be an integer"`
      )
    })
  })
})
