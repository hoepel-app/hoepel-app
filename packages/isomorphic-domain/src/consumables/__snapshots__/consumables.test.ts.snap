// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consumables withConsumableRenamed renames consumable 1`] = `
Consumables {
  "props": Object {
    "consumables": Array [
      Object {
        "name": "Soft drink",
        "priceCents": 50,
      },
      Object {
        "name": "Soup",
        "priceCents": 300,
      },
      Object {
        "name": "Big Chocolate Cookie",
        "priceCents": 250,
      },
    ],
    "tenantId": "my-tenant-id-here",
  },
}
`;
